### PR TITLE
[runtimes] Add a qemu-system executor script

### PR DESCRIPTION
**Pull request by arichardson**
This is useful when trying to test libc++/libc++abi/libunwind against a baremetal enviroment (e.g. picolibc).

See also https://reviews.llvm.org/D155521

### CODE REVIEW

File: libcxx/utils/qemu_baremetal.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python
```
Comment (arichardson) [inline libcxx/utils/qemu_baremetal.py:1]:
```suggestion
#!/usr/bin/env python3
```

**Comment by ldionne** [comment]
I think @mplatings is probably the right person to review this since I have no familiarity with QEMU. However I have no concerns with the patch itself, feel free to merge once @mplatings is happy.

**Review by ldionne — APPROVED** [review]
I am going to accept as libc++ just for the sake of the process.

**Comment by arichardson** [comment]
ping @mplatings

**Comment by mplatings** [comment]
I'd prefer to land https://reviews.llvm.org/D154246 and https://reviews.llvm.org/D155521 as-is since they're well tested. I'm not sure whether this will work in all circumstances since it doesn't support the full interface of `run.py`, and this change can't be exercised by CI.

**Comment by arichardson** [comment]
> I'd prefer to land https://reviews.llvm.org/D154246 and https://reviews.llvm.org/D155521 as-is since they're well tested. I'm not sure whether this will work in all circumstances since it doesn't support the full interface of `run.py`, and this change can't be exercised by CI.

I've run it locally with picolibc so it definitely works, the codesigner interface of run.py is only needed if you use the script for Apple platforms. 

Would you be okay with landing this change, then you could rebase the phabricator reviews on top of this and use q the python script instead of qemu.sh in those reviews? They need to be rebased anyway to get a clean ci run so hopefully this doesn't add too much work?

**Comment by mplatings** [comment]
@domin144 is taking over from me so I hand your question to him...

File: libcxx/utils/qemu_baremetal.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--qemu", type=str, required=True)
+    parser.add_argument("--cpu", type=str, required=False)
+    parser.add_argument("--machine", type=str, default="virt")
+    parser.add_argument(
+        "--qemu-arg", dest="qemu_args", type=str, action="append", default=[]
+    )
+    parser.add_argument(
+        "--semihosting", type=argparse.BooleanOptionalAction, default=True
+    )
```
Comment (domin144) [inline libcxx/utils/qemu_baremetal.py:33]: `BooleanOptionalAction` is a python 3.9 feature. The system image used to run the picolibc is Ubuntu focal, which has python 3.8.

**Review by domin144 — COMMENTED** [review]
Hi @arichardson,

I run the tests from https://reviews.llvm.org/D155521, with this script and it works. I had to do two workarounds though.
- I had to avoid feature from python above 3.8 (see comment inline)
- I had to update picolibc - older one used different numeration of arguments in `argv`

I am happy to rebase and use the new script in the picolibc tests review, when this script is merged.

File: libcxx/utils/qemu_baremetal.py (hunk 2 above)
Comment (arichardson) [reply libcxx/utils/qemu_baremetal.py:33]: Thanks, I'll update to remove that too keep older python compat.

**Pull request merged**